types/persisted-model.d.ts

```diff
@@ -0,0 +1,511 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: loopback-datasource-juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {AnyObject, Options, Callback, PromiseOrVoid} from './common';
+import {ModelBase, ModelData} from './model';
+import {Where, Filter} from './query';
+
+// tslint:disable:no-any
+// tslint:disable:max-line-length
+
+export type PersistedData<
+  T extends PersistedModel = PersistedModel
+> = ModelData<T>;
+
+export interface Count {
+  count: number;
+}
+
+export declare class PersistedModel extends ModelBase {
```

Comment:
Technically speaking, there is no such thing as a `PersistedModel` at juggler level. We have `ModelBase` and `DataAccessObject` mixin that's added to `ModelBase` when a model is attached to a datasource using the default DAO implementation. Similarly, we have `KeyValueAccessObject` mixin that provides API for models connected to key-value stores.

The class `PersistedModel` is implemented in `loopback`, as a thin wrapper providing default DAO methods which throw a helpful message when the model was not attached to a datasource yet.

In the light of that, I feel the currently proposed typedefs are subtly incorrect and I would prefer if we could find a way how to make `.d.ts` files closer to the actual juggler API and behavior.

Thoughts?

Reply:
I'll check

Reply:
LB3 uses mixins at runtime to build model classes. Even though the `PersistedModel` is physically from `loopback` module, we do have the assumption of `PersistedModel` class in many cases in `loopback-datasource-juggler`.

It's hard to describe such contracts in other forms such as interfaces to allow both static and prototype methods/properties.